src/content/post/fastrace-a-modern-approach-to-distributed-tracing-in-rust.mdx

```diff
@@ -0,0 +1,374 @@
+
+While powerful, `tokio-rs/tracing` comes with several significant challenges:
+
+### 1. Ecosystem Fragmentation
+
+By introducing its own logging macros, `tokio-rs/tracing` creates a division with code using the standard `log` crate:
+
+```rust
+// Using log crate
+log::info!("Starting operation");
+
+// Using tracing crate (different syntax)
+tracing::info!("Starting operation");
+```
+
+This fragmentation is particularly problematic for library authors. When creating a library, authors face a difficult choice:
+
+1. Use the `log` crate for compatibility with the broader ecosystem
+2. Use `tokio-rs/tracing` for better observability features
+3. Implement dual support for both systems (increasing maintenance burden)
+
+    This often leads to complex feature-flag systems and wrapper macros:
+
+    ```rust
+    macro_rules! info {
+        ($($arg:tt)*) => {
+            if cfg!(feature = "tracing") {
+                tracing::info!($($arg)*);
+            } else {
+                log::info!($($arg)*);
+            }
+        }
+    }
+    ```
+
+Many libraries choose the first option for simplicity, but miss out on the benefits of tracing. This creates a situation where applications have inconsistent instrumentation across their dependency tree.
+
+### 2. Performance Impact for Libraries
+
+Library authors are particularly sensitive to performance overhead, as their code may be called in tight loops or performance-critical paths. `tokio-rs/tracing`'s overhead can be substantial when enabled, which creates a dilemma:
+
+1. Always enable tracing (and impose overhead on all users)
+2. Don't instrument at all (and lose observability)
+3. Create a complex feature-flag system (increasing maintenance burden)
+
+    This is a common pattern in libraries using `tokio-rs/tracing`:
+
+    ```rust
+    #[cfg_attr(feature = "tracing", tracing::instrument(skip(password), fields(user_id = user.id)))]
+    async fn authenticate(user: &User, password: &str) -> Result<AuthToken, AuthError> {
+        // ...more code...
+    }
+    ```
+
+With `tokio-rs/tracing`, there's no clean way to have zero-cost tracing when disabled. This makes library authors reluctant to add instrumentation to performance-sensitive code paths.
+
+### 3. No Context Propagation
+
+Distributed tracing requires propagating context across service boundaries, but `tokio-rs/tracing` leaves this largely as an exercise for the developer. For example, the [tonic official example](https://github.com/hyperium/tonic/blob/master/examples/src/tracing/server.rs) demonstrates how to trace gRPC service using tonic:
+
+```rust
+// Server setup with `tokio-rs/tracing`
+Server::builder()
+    // Basic span only, no context extraction
+    .trace_fn(|_| tracing::info_span!("grpc_server"))
+    .add_service(MyServiceServer::new(MyService::default()))
+    .serve(addr)
+    .await?;
+```
+
+The above example only creates a basic span but doesn't extract tracing context from the incoming request.
```

Comment:
Can we elaborate a bit here how the configure above would cause a (maybe?) fragmented trace?

As a reader one may think that: so, you can configure tracing in one line, what's the problem here?

Reply:
refined with additional examples